### ISSUE TITLE
Termination tasks with eventlet/gevent worker pool.

### ISSUE THREAD

**Issue by alternativshik** [celery/celery#4019]
This is a feature request about ability to terminate task execution when using eventlet or gevent TaskPool.

Now this feature presents only on Prefork TaskPool, but, I think, this is a useful feature for other Pools.

**Comment by auvipy** [comment]
have you any suggested workarounds?

**Comment by jengelman** [comment]
+1

**Comment by sreecodeslayer** [comment]
I believe the issue am facing while terminating the task has to do with this feature request

Am using the celery [ `v4.1.0` ] with Flask as : `celery -A run.celery worker -P eventlet -c 50 --loglevel=info`

```
[2018-05-11 17:16:28,137: ERROR/MainProcess] pidbox command error: NotImplementedError("<class 'celery.concurrency.eventlet.TaskPool'> does not implement kill_job",)
Traceback (most recent call last):
  File "/home/sreenadh/Projects/test_project/.tp/lib/python3.5/site-packages/kombu/pidbox.py", line 101, in dispatch
    reply = handle(method, arguments)
  File "/home/sreenadh/Projects/test_project/.tp/lib/python3.5/site-packages/kombu/pidbox.py", line 122, in handle_cast
    return self.handle(method, arguments)
  File "/home/sreenadh/Projects/test_project/.tp/lib/python3.5/site-packages/kombu/pidbox.py", line 116, in handle
    return self.handlers[method](self.state, **arguments)
  File "/home/sreenadh/Projects/test_project/.tp/lib/python3.5/site-packages/celery/worker/control.py", line 164, in revoke
    request.terminate(state.consumer.pool, signal=signum)
  File "/home/sreenadh/Projects/test_project/.tp/lib/python3.5/site-packages/celery/worker/request.py", line 248, in terminate
    pool.terminate_job(self.worker_pid, signal)
  File "/home/sreenadh/Projects/test_project/.tp/lib/python3.5/site-packages/celery/concurrency/base.py", line 115, in terminate_job
    '{0} does not implement kill_job'.format(type(self)))
NotImplementedError: <class 'celery.concurrency.eventlet.TaskPool'> does not implement kill_job
```
I believe the killing would be possible using `Eventlet API` corresponding to the `greenthread.kill()` ?

**Comment by MrJeric0** [comment]
+1

**Comment by seiyata** [comment]
+1

**Comment by basitiyidir** [comment]
+1

**Comment by antoarva** [comment]
+1

**Comment by RogerLiNing** [comment]
+1

**Comment by ik4ru52k** [comment]
+1

**Comment by MarcDelahousse** [comment]
+1

**Comment by ioliveros** [comment]
+1

**Comment by auvipy** [comment]
send PR, +1 will be blocked

**Comment by xd404015353** [comment]
+1

**Comment by zhangzhw8** [comment]
+1

**Comment by eugene-s** [comment]
+1

**Comment by remi-debette** [comment]
+1

**Comment by AlphaCaprice** [comment]
+ 1

**Comment by ashish-iycn** [comment]
+1

**Comment by ashishpahwa7** [comment]
+1

**Comment by Alea4jacta6est** [comment]
+

**Comment by auvipy** [comment]
for those commenting +1, https://opencollective.com/celery consider donating to this project.

**Comment by AlphaCaprice** [comment]
Well, this https://celery.readthedocs.io/en/latest/reference/celery.contrib.abortable.html can be useful. You can try to abort your task from outside (set state to ABORTED) and check this state in task with some period, if state is ABORTED you can raise some Exception. What do you think about this?

**Comment by singh-atenekom** [comment]
+1

@AlphaCaprice I tried this approach. I was not able to stop the process from outside. Everything is related gevent third party app.

**Comment by auvipy** [comment]
considering that we are going to move to python3.6+ asyncio/trio can we close this issue?

**Comment by adityamagarde** [comment]
I found a workaround. Might not work for everyone though.

I had a loop that had to run with the heavy process. Couldn't revoke the task so the entire loop ran. So before starting the task, I made a file(let's say _task_id.txt_) and put a logic to check if the file existed on every iteration. I could delete the testfile if I wanted to revoke and raise an exception.

This is similar to AbortableTask as in the current iteration will still run, but was a work around as I couldn't use AbortableTask in my case.

**Comment by thedrow** [comment]
> 
> 
> considering that we are going to move to python3.6+ asyncio/trio can we close this issue as won't fix??

Sadly, yes. No one has fixed this so far so it probably won't get fixed.
If you feel differently, please come up with a PR and we will revisit that decision.

**Comment by oxtime** [comment]
+100

**Comment by Dominux** [comment]
+1

**Comment by adv-zl** [comment]
+1

**Comment by Kanishk-Kumar** [comment]
+1

**Comment by mht2953658596** [comment]
+1

**Comment by mht2953658596** [comment]
+1

**Comment by bengabp** [comment]
Not fixed yet ??

**Comment by Nusnus** [comment]
> Not fixed yet ??

https://github.com/celery/celery/pull/9083

**Comment by bengabp** [comment]
> @bengabp 
> > Not fixed yet ??
> 
> https://github.com/celery/celery/pull/9083

Yeah I still got the error raised even after using terminate=True

**Comment by Nusnus** [comment]
> > @bengabp 
> > > Not fixed yet ??
> > 
> > https://github.com/celery/celery/pull/9083
> 
> Yeah I still got the error raised even after using terminate=True

@ldsink Can you give it a look please 🙏

**Comment by Nusnus** [comment]
@bengabp 
> > @bengabp 
> > > Not fixed yet ??
> > 
> > https://github.com/celery/celery/pull/9083
> 
> Yeah I still got the error raised even after using terminate=True

Can you please share more info?
Are you doing anything differently than what the PR fixes?

**Comment by bengabp** [comment]
No. Just a generic task I have which is running on gevent pool and worker raises not implemented exception when I revoke the task

**Comment by Nusnus** [comment]
> No. Just a generic task I have which is running on gevent pool and worker raises not implemented exception when I revoke the task

Are you using the latest pre-release of Celery?
Just confirming

**Comment by bengabp** [comment]
@Nusnus 
![image](https://github.com/user-attachments/assets/b7719de7-a6d7-4bfb-a22d-1ff5fab4aef6)

**Comment by Nusnus** [comment]
@bengabp 
> @Nusnus 
> ![image](https://github.com/user-attachments/assets/b7719de7-a6d7-4bfb-a22d-1ff5fab4aef6)
> 

This fix was released in 5.5.0b2:
https://github.com/celery/celery/blob/main/Changelog.rst#L193

You will need to upgrade to the latest pre-release or wait for the final release to have the fix from the PR.

**Comment by bengabp** [comment]
Okay
